File: java/src/main/java/br/com/emendes/problems/RotateList.java
package br.com.emendes.problems;

import br.com.emendes.problems.util.ListNode;

/**
 * Dado a head de uma linked list, rotacione a lista para a direta em k vezes.<br><br>
 *
 * Restrições:<br>
 * O número de nodes da lista está no intervalo de [0, 500].<br>
 * -100 <= Node.val <= 100<br>
 * 0 <= k <= 2 * 10⁹
 */
public class RotateList {

  public ListNode rotateRight(ListNode head, int k) {
    if (head == null || head.next == null) return head;

    int listSize = 1;
    ListNode tail = head;

    while(tail.next != null) { // O(n)
      listSize++;
      tail = tail.next;
    }

    k = k % listSize;
    if (k == 0) return head;

    int index = listSize - k - 1;
    ListNode newTail = head;

    while (index > 0) {
      newTail = newTail.next;
      index--;
    }

    tail.next = head;
    ListNode newHead = newTail.next;
    newTail.next = null;

    return newHead;
  }

}

File: javascript/problems/RotateList.js
/**
 * Dado o head de uma linked list, rotacione a lista para a direita em k posições.
 *
 * Restrições:
 *
 * O número de nodes na lista está no intervalo [0, 500].
 *
 * -100 <= Node.val <= 100
 *
 * 0 <= k <= 2 * 10⁹
 *
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
const rotateRight = function (head, k) {
  const length = getLength(head);
  if (length === 0 || length === 1) return head;
  k = k % length;
  if (k === 0) return head;
  let tail = head;
  while (k > 0) {
    tail = tail.next;
    k--;
  }
  let newTail = head;
  while (tail.next) {
    tail = tail.next;
    newTail = newTail.next;
  }
  tail.next = head;
  let newHead = newTail.next;
  newTail.next = null;

  return newHead;
};

const getLength = (head) => {
  let length = 0;
  while (head) {
    length++;
    head = head.next;
  }
  return length;
};

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}
